editor/src/components/canvas/commands/push-intended-bounds-and-update-groups-command.ts

```diff
@@ -0,0 +1,336 @@
+    value: value,
+  }
+}
+
+export const runPushIntendedBoundsAndUpdateGroups = (
+  editor: EditorState,
+  command: PushIntendedBoundsAndUpdateGroups,
+  commandLifecycle: InteractionLifecycle,
+): CommandFunctionResult => {
+  const { statePatch: resizeAncestorsPatch, intendedBounds: extraIndendedBounds } =
+    getGroupUpdateCommands(editor, command)
+
+  const intendedBoundsPatch =
+    commandLifecycle === 'mid-interaction'
+      ? pushCommandStatePatch([...command.value, ...extraIndendedBounds])
+      : []
+
+  return {
+    editorStatePatches: [resizeAncestorsPatch, ...intendedBoundsPatch],
+    commandDescription: `Set Intended Bounds for ${command.value
+      .map((c) => EP.toString(c.target))
+      .join(', ')}`,
+  }
+}
+
+function pushCommandStatePatch(
+  intendedBounds: Array<CanvasFrameAndTarget>,
+): Array<EditorStatePatch> {
+  return [
+    {
+      canvas: {
+        controls: {
+          strategyIntendedBounds: { $push: intendedBounds },
+        },
+      },
+    },
+  ]
+}
+
+function getGroupUpdateCommands(
+  editor: EditorState,
+  command: PushIntendedBoundsAndUpdateGroups,
+): { statePatch: EditorStatePatch; intendedBounds: Array<CanvasFrameAndTarget> } {
+  const targets = command.value
+
+  // we are going to mutate this as we iterate over targets
+  let updatedGlobalFrames: { [path: string]: CanvasFrameAndTarget | undefined } = {}
+
+  function getGlobalFrame(path: ElementPath): CanvasRectangle {
+    return forceNotNull(
+      `Invariant: found null globalFrame for ${EP.toString(path)}`,
+      updatedGlobalFrames[EP.toString(path)]?.frame ??
+        nullIfInfinity(
+          MetadataUtils.findElementByElementPath(editor.jsxMetadata, path)?.globalFrame,
+        ),
+    )
+  }
+
+  targets.forEach((frameAndTarget) => {
+    const parentPath = EP.parentPath(frameAndTarget.target)
+    const parentIsGroup = treatElementAsGroupLike(editor.jsxMetadata, parentPath)
+
+    if (!parentIsGroup || parentPath == null) {
+      // bail out
+      return
+    }
+
+    // I assume that affectedAncestors are ordered bottom-up
+    // the ancestor's globalFrame shall be the union of the current global frame and the target's frame
+    const childrenExceptTheTarget = MetadataUtils.getChildrenPathsOrdered(
+      editor.jsxMetadata,
+      editor.elementPathTree,
+      parentPath,
+    ).filter((c) => !EP.pathsEqual(c, frameAndTarget.target))
+    const childrenGlobalFrames = childrenExceptTheTarget.map(getGlobalFrame)
+
+    const newGlobalFrame = boundingRectangleArray([...childrenGlobalFrames, frameAndTarget.frame])
+    if (newGlobalFrame != null) {
+      updatedGlobalFrames[EP.toString(parentPath)] = {
+        frame: newGlobalFrame,
+        target: parentPath,
+      }
+    }
+  })
+
+  // we REALLY need a lens for this!
+  let commandsToRun: Array<CanvasCommand> = []
+
+  // okay so now we have a bunch of new globalFrames. what do we do with them?
+  Object.keys(updatedGlobalFrames).forEach((pathStr) => {
```

Comment:
[`Object.entries`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) is your friend